shelley/chain-and-ledger/shelley-spec-ledger-test/test/Test/Control/Iterate/SetAlgebra.hs

```diff
@@ -410,5 +416,236 @@ threeWay delegs stake hk =
 threeWayTest :: TestTree
 threeWayTest = testProperty "eval-materialize-intersectDom" threeWay
 
+
+-- ================================================
+-- slow property tests
+-- ================================================
+
+newtype Key = Key Int
+  deriving (Eq,Ord)
+
+instance Show Key where
+   show (Key n) = "k"++show n
+
+-- ----------------------------
+newtype Range = Range Int
+  deriving (Eq,Ord,Num)
+
+instance Show Range where
+  show (Range n) = show n
+
+instance Semigroup Range where (Range x) <> (Range y) = Range (x + y)
+instance Monoid Range where mempty = Range 0
+
+
+-- ===========================================================
+flip_rng :: (Ord b, Num b) => List a b -> List b b
+flip_rng (UnSafeList xs) = fromPairs (+) (map (\ (a,b) -> (b,b)) xs)
+
+bimap:: (Ord k,Ord v) => Map k v -> BiMap v k v
+bimap xs = biMapFromList (\ old new -> new) (Map.toList xs)
+
+duplicate :: Ord a => Set.Set a -> Map.Map a a
+duplicate s = foldr (\ a m -> Map.insert a a m) Map.empty s
+
+btest :: Exp Bool -> Property
+btest exp = (compute exp) === (runBoolExp exp)
+
+qtest :: (Ord key,Eq(f key a),Show(f key a)) => Exp (f key a) -> Property
+qtest exp =  (compute exp) === (run (compile exp))
+
+-- ======================================================
+
+slowFastEquiv :: TestTree
+slowFastEquiv = testGroup "slowFastEquiv" (map f many)
+  where f (prop,name) = testProperty name prop
+
+
+type STest =
+   Key ->             -- k
+   Range ->           -- v
+   Map Key Range ->   -- m1
+   Map Key Range ->   -- m2
+   Set.Set Key ->     -- s1
+   Set.Set Key ->     -- s2
+   Set.Set Range ->   -- rs
+   List Key Range ->  -- ls
+   Property
+
+many :: [(STest,String)]
```

Comment:
set algebra tests?

Reply:
oh I see, you manipulate this list into the thing above. Is this style perhaps easier to read: https://github.com/input-output-hk/cardano-ledger-specs/blob/master/shelley/chain-and-ledger/shelley-spec-ledger-test/test/Test/Shelley/Spec/Ledger/Serialisation/Tripping/CBOR.hs#L161

Reply:
Everythign is wrapped into one bit test at the end of the file.